src/platforms/hosted/jlink.c

```diff
@@ -277,6 +234,313 @@ static bool jlink_claim_interface(void)
+	if (jlink.interface_frequency[interface].base != 0U)
+		return true;
+
+	/* Get the selected interface */
+	const uint8_t selected_interface = jlink_selected_interface();
+
+	if (selected_interface != interface) {
+		/* If the selected interface doesn't match the requested interface, select it, let's hope this doesn't mess something up elsewhere */
+		DEBUG_WARN("Trying to get frequency for interface %s but it is not selected, selecting it\n",
+			jlink_interface_to_string(interface));
+
+		/* select the requested interface */
+		if (!jlink_select_interface(interface))
+			return false;
+	}
+
+	/* Get the frequency info for the selected interface */
+	uint8_t buffer[6U];
+	if (!jlink_simple_query(JLINK_CMD_INTERFACE_GET_BASE_FREQUENCY, buffer, sizeof(buffer)))
+		return false;
+
+	struct jlink_interface_frequency *const interface_frequency = &jlink.interface_frequency[interface];
+
+	interface_frequency->base = read_le4(buffer, JLINK_INTERFACE_BASE_FREQUENCY_OFFSET);
+	interface_frequency->min_divisor = read_le2(buffer, JLINK_INTERFACE_MIN_DIV_OFFSET);
+
+	/* This is an assumption, if the J-Link was configured before we started, this may not be true, but we have no way to know */
+	interface_frequency->current_divisor = interface_frequency->min_divisor;
+
+	DEBUG_INFO("%s interface frequency:\n\tBase frequency: %uHz\n\tMinimum divisor: %u\n",
+		jlink_interface_to_string(interface), interface_frequency->base, interface_frequency->min_divisor);
+
+	// Restoring the selected interface seemed to cause issues sometimes (unsure if culprit)
+	// so we don't do it for now, it didn't seem to be necessary, saving in case it proves to be necessary later
+	// if (selected_interface != interface) {
+	// 	/* If the selected interface didn't match the requested interface, restore the selected interface */
+	// 	DEBUG_WARN("Restoring selected interface to %s\n", jlink_interface_to_string(selected_interface));
+
+	// 	/* select the requested interface, we don't consider this a critical failure, as we got what we wanted */
+	// 	if (!jlink_select_interface(selected_interface))
+	// 		DEBUG_ERROR("Failed to restore selected interface to %s\n", jlink_interface_to_string(selected_interface));
+	// }
+
+	return true;
+}
+
+bool jlink_set_interface_frequency(const uint8_t interface, const uint32_t frequency)
+{
+	if (!(jlink.capabilities & JLINK_CAPABILITY_INTERFACE_FREQUENCY)) {
+		DEBUG_WARN("J-Link does not support interface frequency command\n");
+		return false;
+	}
+
+	/* Get the selected interface */
+	const uint8_t selected_interface = jlink_selected_interface();
+
+	if (selected_interface != interface) {
+		/* If the selected interface doesn't match the requested interface, select it, let's hope this doesn't mess something up elsewhere */
+		DEBUG_WARN("Trying to set frequency for interface %s but it is not selected, selecting it\n",
```

Comment:
We think this makes a touch more sense as a DEBUG_INFO() statement as it's not so much a warning but informational

Reply:
I went for warning because I considered this to be a potentially problematic action, hence the commented out "restoring" code included, so far it didn't prove to be an issue so perhaps you're right

Reply:
Ahhh, that makes sense. Sure, this can be left as a warning then.